geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/ServerConnection.java

```diff
@@ -254,6 +254,8 @@ public static void emptyCommBufferPool() {
   @MutableForTesting
   private static boolean TEST_VERSION_AFTER_HANDSHAKE_FLAG;
 
+  private NioSslEngine sslEngine;
```

Comment:
We need to strive to make fields like this `final`. There are several places in the new code where you close the `NioSslEngine` and set the field back to null. There are also lots of null checks. This makes the code thread unsafe because there's no form of synchronization or use of volatile. Non-final fields tend to lower performance and make the code a lot busier with null checks.

Why do you really need this to be mutable? And what is the state change for that changes this from non-null to null? Why does the `ServerConnection` remain valid and usable after closing and nulling out the `NioSslEngine`?

There are two possible approaches I would take:

1) I would try very hard to make `NioSslEngine sslEngine` a final field. 

I might even change it to be an interface so that I can either pass in a `NioSslEngine` or a dummy implementation of the interface to the constructor. A dummy implementation does nothing or returns values from methods that causes the calling code to not use nio ssl. 

2) If I really need the field to be mutable AND the class instance is exposed more than one thread, then I would use an AtomicReference to make sure it's thread-safe:
```
private final AtomicReference<NioSslEngine> sslEngine = new AtomicReference<>();
```
Using an interface with a dummy implementation instead of null can also be used with AtomicReferences.